development/cpp/internal_rendering_architecture.rst

```diff
@@ -0,0 +1,694 @@
+
+A 2D signed distance field representing the viewport is automatically generated
+if an user shader requests it. This can be used for various effects in custom
+shaders, such as 2D global illumination.
+
+**2D SDF generation GLSL shader:**
+
+- `https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/canvas_sdf.glsl <servers/rendering/renderer_rd/shaders/canvas_sdf.glsl>`__
+
+3D rendering techniques
+-----------------------
+
+Batching and instancing
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In the clustered and mobile backends, Vulkan instancing is used to group
+rendering of identical objects for performance. This is not as fast as static
+mesh merging, but it still allows instances to be culled individually.
+
+In the compatibility backend, a more limited form of 3D instancing is also used.
+**TODO:** Check if this is the case?
+
+Light, decal and reflection probe rendering
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. note::
+
+  Reflection probe and decal rendering are currently not available in the
+  Compatibility backend.
+
+As its name implies, the Clustered Forward backend uses clustered lighting. This
+allows using as many lights as you want; performance largely depends on screen
+coverage. Shadow-less lights can be almost free if they don't occupy much space
+on screen.
+
+The Clustered Forward backend also supports rendering several directional lights
+at the same time (albeit with lower shadow quality when more than one has
+shadows enabled). Other backends are limited to only one DirectionalLight
+rendered at a given time.
+
+The Forward Mobile backend uses a single-pass lighting approach, with a
+limitation of 8 OmniLights + 8 SpotLights affecting each Mesh *resource*. This
+limit is hardcoded and can't be adjusted in the project settings.
+
+The Compatibility backend uses a hybrid single-pass + multi-pass lighting
+approach. Lights without shadows are rendered in a single pass (using less
+accurate additive blending for performance reasons). Lights with shadows are
+rendered in multiple passes. This is required for compatibility with low-end
+mobile devices, but it has a huge cost to performance. As a result, it's
+recommended to only have a handful of lights with shadows in the camera frustum
+at most. The maximum number of lights visible at once can be adjusted in the
+project settings.
+
+In all 3 backends, lights without shadows are much cheaper than lights with shadows.
+Godot currently doesn't separate static shadow rendering from dynamic shadow
+rendering, but this is planned in a future release.
+
+Clustering is also used for reflection probe and decal rendering in the
+Clustered Forward backend.
+
+Shadow mapping
+^^^^^^^^^^^^^^
+
+Both Clustered Forward and Forward Mobile backends use the same shadow mapping
+approach, with the use of a Vogel disk for improved filtering. Dithering is also
+used to improve shadow appearance at the cost of some visible graininess.
```

Comment:
```suggestion
Both Clustered Forward and Forward Mobile backends use percentage closer 
filtering (PCF) to filter shadow maps and create a soft penumbra. Instead of
using a fixed PCF pattern, they use a vogel disk pattern which allows for changing
the number of samples and smoothly changing the quality.
```